### PR TITLE
Add more string functions

### PR DESCRIPTION
**Pull request by HactarCE**
Done in this PR:

- [x] `ARRAYTOTEXT`
- [x] `CONCATENATE`
- [x] `CONCAT` (already done)
- [x] `LEN`/`LENB` (already done)
- [x] `CHAR`
- [x] `UNICHAR`
- [x] `CODE`
- [x] `UNICODE`
- [x] `EXACT`
- [x] `LOWER`
- [x] `UPPER`
- [x] `PROPER`
- [x] `NUMBERVALUE`
- [x] `T`
- [x] `TRIM`
- [x] `LEFT`/`LEFTB`
- [x] `MID`/`MIDB`
- [x] `RIGHT`/`RIGHTB`

Leaving for future PR(s):

- [ ] `TEXT`
- [ ] `VALUE`
- [ ] `VALUETOTEXT`
- [ ] `TEXTAFTER`
- [ ] `TEXTBEFORE`
- [ ] `TEXTJOIN`
- [ ] `TEXTSPLIT`
- [ ] `FIND`/`FINDB`
- [ ] `SEARCH`/`SEARCHB`
- [ ] `REPLACE`/`REPLACEB`
- [ ] `SUBSTITUTE`

My linguistics minor didn't prepare me to understand what the [`PHONETIC`](https://support.microsoft.com/en-us/office/phonetic-function-9a329dac-0c0f-42f8-9a55-639086988554) function does, and apparently Google Sheets feels the same way (they don't implement it) so let's skip it for now.

### CODE REVIEW

**Comment by luke-quadratic** [comment]
Approved on functionality - all formulas align with other standard spreadsheet approaches 👍

File: quadratic-core/src/formulas/functions/string.rs [hunk 1]
```diff
@@ -10,18 +10,212 @@ pub const CATEGORY: FormulaFunctionCategory = FormulaFunctionCategory {
...
+                        return Err(RunErrorMsg::InvalidArgument.with_span(span))
+                    }
+                }
+            }
+        ),
+        formula_fn!(
+            /// Same as `CONCAT`, but kept for compatibility.
+            #[examples("CONCATENATE(\"Hello, \", C0, \"!\")")]
+            fn CONCATENATE(strings: (Iter<String>)) {
+                strings.try_fold(String::new(), |a, b| Ok(a + &b?))
```
Comment (ddimaria) [inline quadratic-core/src/formulas/functions/string.rs:44]: Just call CONCAT?

File: quadratic-core/src/formulas/functions/string.rs [hunk 2]
```diff
@@ -45,16 +239,320 @@ fn get_functions() -> Vec<FormulaFunction> {
...
+            ///
+            /// [Other forms of whitespace][whitespace], including tabs and
+            /// newlines, are preserved.
+            ///
+            /// [whitespace]: https://en.wikipedia.org/wiki/Whitespace_character
+            #[examples("TRIM(\"    a    b    c    \")=\"a b c\"")]
+            #[zip_map]
+            fn TRIM([s]: String) {
+                // This would be a one-liner if we wanted to apply it to all
+                // whitespace: `s.split_whitespace().join(" ")`
```
Comment (ddimaria) [inline quadratic-core/src/formulas/functions/string.rs:310]: 👍

File: quadratic-core/src/formulas/functions/string.rs [hunk 3]
```diff
@@ -45,16 +239,320 @@ fn get_functions() -> Vec<FormulaFunction> {
...
+        ),
+        formula_fn!(
+            /// Capitalizes letters that do not have another letter before them,
+            /// and lowercases the rest.
+            #[examples("PROPER(\"ὈΔΥΣΣΕΎΣ is my FAVORITE character!\") = \"Ὀδυσσεύς Is My Favorite Character!\"")]
+            #[zip_map]
+            fn PROPER([s]: String) {
+                let mut last_char = '\0';
+                let mut ret = String::new();
+                // Convert to lowercase first so that we get correct handling of
```
Comment (ddimaria) [inline quadratic-core/src/formulas/functions/string.rs:347]: Great documentation 💯

File: quadratic-core/src/formulas/functions/string.rs [hunk 4]
```diff
@@ -45,16 +239,320 @@ fn get_functions() -> Vec<FormulaFunction> {
...
+                    actual_number_string
+                        .parse::<f64>()
+                        .map_err(|_| RunErrorMsg::InvalidArgument.with_span(s.span))
+                }
+            }
+        ),
+        // Comparison
+        formula_fn!(
+            /// Returns whether two strings are exactly equal, using
+            /// case-sensitive comparison (but ignoring formatting).
```
Comment (ddimaria) [inline quadratic-core/src/formulas/functions/string.rs:458]: What is `formatting` in this context?

**Review by ddimaria — APPROVED** [review]
Terse, readable 🙌  and well-written code, thorough user-facing and code-level documentation 🚀

**Pull request merged**